eth_abi/registry.py

```diff
@@ -0,0 +1,131 @@
+from cytoolz import curry
+
+from .utils.parsing import collapse_type
+
+
+class ABIRegistryError(Exception):
+    pass
+
+
+class NoEntryFound(ABIRegistryError):
+    pass
+
+
+class PredicateMapping:
+    """
+    Acts as a mapping from predicate functions to values.  Values are retrieved when their
+    corresponding predicate matches a set of given inputs.  Predicates can also be labeled to
+    facilitate removal from the mapping.
+    """
+    def __init__(self):
+        self._values = {}
+        self._labeled_predicates = {}
+
+    def add(self, predicate, value, label=None):
+        if predicate in self._values:
+            raise ValueError('Predicate already exists')
+
+        if label is not None:
+            if label in self._labeled_predicates:
+                raise ValueError('Predicate with label "{}" already exists'.format(label))
+
+            self._labeled_predicates[label] = predicate
+
+        self._values[predicate] = value
+
+    def find(self, *args, **kwargs):
```

Comment:
Is the `*args, **kwargs` here because you aren't yet sure what the arguments will be, or to make this *generic*?  I think I'm ok with us making this implementation specific to whatever we *know* the arguments to our `predicate` functions are going to be so that we can eliminate one axis of validation/errors that this is able to produce by forcing things to fail at the `find(..)` invocation rather than somewhere deeper in the `predicate(...)` invocation.

Reply:
It was to make it generic but we can make it specific to its intended use.  I'll change it.